WORKSPACE

```diff
@@ -40,6 +40,35 @@ load("//specs2:specs2_junit.bzl", "specs2_junit_repositories")
 
 specs2_junit_repositories()
 
+load("//scala/scalafmt:scalafmt.bzl", "scalafmt_default_config", "scalafmt_repositories")
+
+scalafmt_repositories()
+
+scalafmt_default_config()
+
+RULES_JVM_EXTERNAL_TAG = "3.1"
+
+http_archive(
```

Comment:
I don't know why I can't import `org.scalameta:scalafmt-core_2.11:2.0.0` by `scala_maven_import_external`, but `rules_jvm_external` works.

Reply:
```
ERROR: /home/borkaehw/Desktop/code/rules_scala/test/scalafmt/BUILD:9:1: ScalaFmt test/scalafmt/test/scalafmt/formatted.scala failed: Worker process did not return a WorkResponse:

---8<---8<--- Start of log, file at /home/borkaehw/.cache/bazel/_bazel_borkaehw/5e84e9463749ae9195051aff1f87cebb/bazel-workers/worker-1-ScalaFmt.log ---8<---8<---
Exception in thread "main" java.lang.NoClassDefFoundError: pprint/TPrintColors
	at org.scalafmt.config.Config$.fromHoconFile$default$3(Config.scala:46)
	at io.bazel.rules_scala.scalafmt.ScalafmtProcessor.processRequest(ScalafmtRunner.scala:40)
	at io.bazel.rulesscala.worker.GenericWorker.runPersistentWorker(GenericWorker.java:45)
	at io.bazel.rulesscala.worker.GenericWorker.run(GenericWorker.java:111)
	at io.bazel.rules_scala.scalafmt.ScalafmtRunner$.main(ScalafmtRunner.scala:16)
	at io.bazel.rules_scala.scalafmt.ScalafmtRunner.main(ScalafmtRunner.scala)
Caused by: java.lang.ClassNotFoundException: pprint.TPrintColors
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:583)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
	... 6 more
---8<---8<--- End of log ---8<---8<---
```

Reply:
this is because scalafmt has dependencies.
scala_maven_import_external uses the bazel (and rules_scala) idiom of statically listing the dependencies.
rules_jvm_external does this implicitly (and in build time given you didn't check in a _install.json file).
see this [gist](https://gist.github.com/ittaiz/3655a8b65ec61460a4bbdf38932b8c75) for 90% of what you need

Reply:
I've given it some more thought and I'd really prefer not adding another "deps" pattern or rules_external. The way to go is toolchains. We need to have a scalafmt toolchain that defines it's dependencies. We need to offer one out of the box (maybe two also for 2.12) and users can always build and configure their own toolchain with their own deps.
Having said that I'm willing to let this be a followup PR if you'd rather make this one more concise (but still not with rules_jvm_external). WDYT?

Reply:
It fixes it, thanks! I am curious how do you generate these dependencies?

Reply:
We have a tool similar to bazel-deps that does this. It's similar to bazel-deps but different in a few key ways (it has support for a bazel repo with managed dependencies for a many repo org, it lays out the deps in external repositories and not build targets so they only get built if needed and a few other things).
We're actually planning to open-source it since we're open sourcing a few of our tools and they're using this pattern so we'll need this in.
cc @natansil as the one driving the requirement to open source it and cc @or-shachar as the one who'll probably own the open sourcing in some capacity

Reply:
The tool sounds amazing. I wonder why don't we just use `rules_jvm_external`?

Reply:
1. It’s a long discussion which I’d rather not do on this PR.
2. For that matter you can also ask why not bazel-deps.
3. I think this challenge hasn’t hit us so far because we strive to have minimum dependencies and so making it easier to have large transitive graphs is actually not that great. 
4. I think that given that everything works via toolchains then it will be much easier for users to override whatever rules scala suggests out of the box and so at that point maybe introducing an easier entry level tool while still allowing more advanced users to do whatever they want is a better trade off.